File: CODE/Dixon_Price.f90
!=========================================
! Dixon & Price function
!
! n= variabile (25 (default))
!        
! valore ottimo della funzione obiettivo = 0.0
!
!         -10.0 =< x(i) =< 10.0   
!====================================================    

SUBROUTINE FUNCT(N,X,F)

      IMPLICIT NONE

      INTEGER          :: N,I
      DOUBLE PRECISION :: X(N),F

	  F=0.D0
      DO I = 2,N
         F = F+dble(I)*(2.D0*x(I)**2-x(I-1))**2    
      ENDDO
      F = F+(x(1)-1.D0)**2

      RETURN

END

SUBROUTINE DATA(N,L,U,FOTT)

      IMPLICIT NONE

      INTEGER          :: N
      DOUBLE PRECISION :: L(N),U(N),FOTT
      INTEGER          :: I

      DO I=1,N
         L(I) = -10.d0
         U(I) =  10.d0
      ENDDO

      FOTT = 0.D0

      RETURN

END


subroutine grad(n,x,g)


implicit none

integer				::n, i
double precision	::x(n), g(n)

do i=1,n

if (i==1) then

	g(1) = 2.0d0*(x(i)-1.0d0)-2.0d0*dble(i+1)*(2.0d0*(x(2)**2)-x(1))
else
	if (i< n) then
		
		g(i) = 8.0d0*dble(i)*x(i)*(2.0d0*(x(i)**2)-x(i-1))-2.0d0*dble(i+1)*(2.0d0*(x(i+1)**2)-x(i)) 

	else
		
		g(i) = 8.0d0*dble(i)*x(i)*(2.0d0*(x(i)**2)-x(i-1))

	endif
endif

enddo	

return

end subroutine grad

File: Python/dixon_price.py
#=========================================
# Dixon & Price function
#
# n= variabile (25 (default))
#        
# valore ottimo della funzione obiettivo = 0.0
#
#         -10.0 =< x(i) =< 10.0   
#====================================================    

import numpy as np

def funct(x):
    n = len(x)
    f = 0.0

    for i in range(2, n+1):
        f += float(i) * (2.0 * x[i-1]**2 - x[i-2])**2

    f += (x[0] - 1.0)**2

    return f

def data(n):
    l = np.full(n, -10.0)
    u = np.full(n, 10.0)
    fott = 0.0

    return l, u, fott

def grad(n, x):
    g = np.zeros(n)

    for i in range(1, n+1):
        if i == 1:
            g[0] = 2.0 * (x[0] - 1.0) - 2.0 * float(i + 1) * (2.0 * x[1]**2 - x[0])
        elif i < n:
            g[i-1] = 8.0 * float(i) * x[i-1] * (2.0 * x[i-1]**2 - x[i-2]) - 2.0 * float(i + 1) * (2.0 * x[i]**2 - x[i-1])
        else:
            g[i-1] = 8.0 * float(i) * x[i-1] * (2.0 * x[i-1]**2 - x[i-2])

    return g


if __name__ == "__main__":
    # Example Usage
    n_value = 25
    x_values = np.random.uniform(-10.0, 10.0, n_value)  # Replace with your actual input values

    result_funct = funct(x_values)
    result_data = data(n_value)
    result_grad = grad(n_value, x_values)

    print("Result (Function):", result_funct)
    print("Result (Data):", result_data)
    print("Result (Gradient):", result_grad)
